unity-client/Assets/Scripts/MainScripts/DCL/Controllers/HUD/BuilderProjectsPanel/Scripts/ScenesController/ScenesViewController.cs

```diff
@@ -0,0 +1,201 @@
+﻿using System;
+using System.Collections.Generic;
+using Object = UnityEngine.Object;
+
+/// <summary>
+/// This class is responsible for receiving a list of scenes and merge it with the previous list
+/// discriminating deployed and project scenes and triggering events when a new set of scenes arrive or
+/// when new scenes are added or removed.
+/// It instantiate and hold all the SceneCardViews to make them re-utilizable in every menu section screen.
+/// </summary>
+internal class ScenesViewController : IDisposable
+{
+    public event Action<Dictionary<string, SceneCardView>> OnDeployedScenesSet;
+    public event Action<SceneCardView> OnDeployedSceneAdded;
+    public event Action<SceneCardView> OnDeployedSceneRemoved;
+    public event Action<Dictionary<string, SceneCardView>> OnProjectScenesSet;
+    public event Action<SceneCardView> OnProjectSceneAdded;
+    public event Action<SceneCardView> OnProjectSceneRemoved;
+
+    public Dictionary<string, SceneCardView> deployedScenes { private set; get; } = new Dictionary<string, SceneCardView>();
+    public Dictionary<string, SceneCardView> projectScenes { private set; get; } = new Dictionary<string, SceneCardView>();
+
+    private readonly ScenesRefreshHelper scenesRefreshHelper = new ScenesRefreshHelper();
+    private readonly SceneCardView sceneCardViewPrefab;
+
+    /// <summary>
+    /// Ctor
+    /// </summary>
+    /// <param name="sceneCardViewPrefab">prefab for scene's card</param>
+    public ScenesViewController(SceneCardView sceneCardViewPrefab)
+    {
+        this.sceneCardViewPrefab = sceneCardViewPrefab;
+    }
+
+    /// <summary>
+    /// Set current user scenes (deployed and projects)
+    /// </summary>
+    /// <param name="scenesData">list of scenes</param>
+    public void SetScenes(List<ISceneData> scenesData)
+    {
+        scenesRefreshHelper.Set(deployedScenes, projectScenes);
+
+        deployedScenes = new Dictionary<string, SceneCardView>();
+        projectScenes = new Dictionary<string, SceneCardView>();
+
+        // update or create new scenes view
+        for (int i = 0; i < scenesData.Count; i++)
+        {
+            SetScene(scenesData[i]);
+        }
+
+        // remove old deployed scenes
+        if (scenesRefreshHelper.oldDeployedScenes != null)
+        {
+            using (var iterator = scenesRefreshHelper.oldDeployedScenes.GetEnumerator())
+            {
+                while (iterator.MoveNext())
+                {
+                    OnDeployedSceneRemoved?.Invoke(iterator.Current.Value);
+                    DestroyCardView(iterator.Current.Value);
```

Comment:
Shouldn't we destroy the card before calling the event.

Reply:
I personally prefer calling the event before destroying the game object in case some subscribed class need to do something with that data before it's destroyed, kind like unity does calling OnDestroy just before actually destroying the game object.
Did this arise any big concern or do you think that there might be an issue doing it this way?

Reply:
No, no. I was just wondering if it was a mistake. if it is intended, let's keep it!